Document how to regenerate the `configure` script

MisterDA opened the pull request:
On distributions with a different version of GNU Autoconf than the one that is checked by the CI, the generated `configure` script might differ. Suggest using Docker and `git rebase --exec`.

No change entry needed.

dustanddreams commented on HACKING.adoc:
```diff
@@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
+`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a
+commit alter the `configure` script, the hook also checks that
+committed `configure` script is up-to-date. The `configure` script may
+be regenerated by running `make -B configure`, with GNU Autoconf
```
Isn't `tools/autogen` preferred to `make configure`? (See e.g. `VERSION`)

MisterDA replied:
The CI uses `make -B configure` and the Makefile has:

```makefile
configure: tools/autogen configure.ac aclocal.m4 build-aux/ocaml_version.m4
	$<
```

MisterDA replied:
and the pre-commit hook says:
> `configure.ac` does not appear to generate configure
> Try running `make -B configure` and stage the changes

dustanddreams replied:
Then it would be nice to update `VERSION` to mention `make -B configure` as well, for consistency.

dustanddreams commented on HACKING.adoc:
```diff
@@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
+
+If you're using a version of GNU Autoconf that differs from our CI,
+then the generated `configure` script might differ too. The latest
+Ubuntu LTS packages may be used as a reasonnable base to generate this
```
```suggestion
Ubuntu LTS packages may be used as a reasonable base to generate this
```

shindere commented:
Miod Vallat (2024/11/14 05:35 -0800):
> @dustanddreams commented on this pull request.
>
>
>
> > @@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
>
>  You can also setup a git commit-hook to automatically run `check-typo`
>  on the changes you commit, by copying the file
> -`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a commit
> -alter the `configure` script, the hook also checks that committed `configure`
> -script is up-to-date.
> +`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a
> +commit alter the `configure` script, the hook also checks that
> +committed `configure` script is up-to-date. The `configure` script may
> +be regenerated by running `make -B configure`, with GNU Autoconf
>
> Isn't `tools/autogen` preferred to `make configure`? (See e.g.
> `VERSION`)

It's definitely my own preference indeed, FWIW.

The trick is that, in theory, the `Makefile` is not supposed to exist at
configure-generation time. We'dmore have just a `Makefile.in` which
would be instanciated ito a `Makefile` by configure itself, so `make
configure` is a bit a chicken-and-egg thing, to me.

shindere commented:
Shouldn't we be more explicit about which version of Autoconf we use? At
the risk/cost of course of forgetting to update this... :(

shindere commented:
Antonin Décimo (2024/11/14 07:41 -0800):
> @MisterDA commented on this pull request.
>
>
>
> > @@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
>
>  You can also setup a git commit-hook to automatically run `check-typo`
>  on the changes you commit, by copying the file
> -`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a commit
> -alter the `configure` script, the hook also checks that committed `configure`
> -script is up-to-date.
> +`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a
> +commit alter the `configure` script, the hook also checks that
> +committed `configure` script is up-to-date. The `configure` script may
> +be regenerated by running `make -B configure`, with GNU Autoconf
>
> The CI uses `make -B configure` and the Makefile has:
>
> ```makefile
> configure: tools/autogen configure.ac aclocal.m4 build-aux/ocaml_version.m4
> 	$<
> ```

Yeah but see my explanation on why I think that's not so much the right
way to do things. More obectively: I don't htink calling autoconf
through make is the most sustainable way of doing things, nor is it
super GNUish.

shindere commented:
Antonin Décimo (2024/11/20 04:48 -0800):
> @MisterDA commented on this pull request.
>
>
>
> > @@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
>
>  You can also setup a git commit-hook to automatically run `check-typo`
>  on the changes you commit, by copying the file
> -`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a commit
> -alter the `configure` script, the hook also checks that committed `configure`
> -script is up-to-date.
> +`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a
> +commit alter the `configure` script, the hook also checks that
> +committed `configure` script is up-to-date. The `configure` script may
> +be regenerated by running `make -B configure`, with GNU Autoconf
>
> and the pre-commit hook says:
> > `configure.ac` does not appear to generate configure
> > Try running `make -B configure` and stage the changes

I'd suggest to change this, too.

shindere commented:
Miod Vallat (2024/11/20 04:51 -0800):
> @dustanddreams commented on this pull request.
>
>
>
> > @@ -575,14 +575,39 @@ runs much faster than a full `./tools/check-typo`, typically instantly.
>
>  You can also setup a git commit-hook to automatically run `check-typo`
>  on the changes you commit, by copying the file
> -`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a commit
> -alter the `configure` script, the hook also checks that committed `configure`
> -script is up-to-date.
> +`tools/pre-commit-githook` to `.git/hooks/pre-commit`. If changes in a
> +commit alter the `configure` script, the hook also checks that
> +committed `configure` script is up-to-date. The `configure` script may
> +be regenerated by running `make -B configure`, with GNU Autoconf
>
> Then it would be nice to update `VERSION` to mention `make -B
> configure` as well, for consistency.

Please no. The other way around.

MisterDA committed:
Suggest regenerating configure by running tools/autogen directly

MisterDA commented:
I've changed the docs and scripts to always use `tools/autogen` over `make -B configure`. The `Makefile` keeps the rule to regenerate the `configure` script, for compatibility.

> Shouldn't we be more explicit about which version of Autoconf we use?

The `AC_PREREQ` macro that we use, and the `check-typo` script kind of already cover that.

dustanddreams reviewed (approved):
I don't like the `Autoconf` spelling (with capital `A`), but that's the name of the software (not the command), so why not.

Thanks for making sure all instructions regenerating `configure` are now consistent.

shindere commented:
Miod Vallat (2024/11/20 10:53 -0800):
> Thanks for making sure all instructions regenerating `configure` are
> now consistent.

I join my thanks to those of @dustanddreams.

shindere commented:
Antonin Décimo (2024/11/20 10:29 -0800):
> I've changed the docs and scripts to always use `tools/autogen` over
> `make -B configure`.

Plenty of thanks for having done so, @MisterDA.

> The `Makefile` keeps the rule to regenerate the `configure` script,
> for compatibility.

Yes, that's very good too, indeed.

> > Shouldn't we be more explicit about which version of Autoconf we use?
>
> The `AC_PREREQ` macro that we use, and the `check-typo` script kind of
> already cover that.

You are right.

shindere commented:
I tend to agree with @dustanddreams in preferring autoconf (lowercase)
to Autoconf (uppercase). Especially in ethe output of commands,
because both the command and the package name are in lowercase.

perhaps we could keep the uppercase iffwe write `GNU Autoconf` and stick
to lower case when autoconf appears alone / in a context that
suggests the command / package name?

Also, although it's not part of this PR strictly, in the first
commit you touch a part of `HACKING.adoc` which says:

`the hook also checks that committed `configure` script`

How about say *the* committed `configure` script?

MisterDA committed:
Suggest a Docker image to regenerate the configure script

In case the user's system and the current CI ship a different version
of GNU Autoconf, the user-generated configure s cript might be
rejected by the CI. Suggest a Docker image based on the latest Ubuntu
LTS, which should match the CI and help .

MisterDA commented:
I've kept the lowercase spelling everywhere as before, and fixed the typo.
Thanks for the reviews.

shindere commented:
Thanks a lot for your patience with your picky reviewer, Antonin.

I approved the PR but the Higiene check complains now. Can you make it
happy again please?

Happy to merge then.

MisterDA commented:
> I approved the PR but the Hygiene check complains now. Can you make it happy again please?

It's just missing a changes entry… I'd like to give credit to the picky reviewers, but is an entry really needed for this PR?

shindere commented:
No no, apologize, I shouldhave added the label. I'll do that right now
and merge.
